### DISCUSSION TITLE
Handling Result On Success Run of a ValueTask

### DISCUSSION THREAD

**Discussion by stefanoIT** [TheCodeTraveler/AsyncAwaitBestPractices#200] [Ideas]
Hello @brminnick and Everyone. 
I may find it useful to add a callback Action to handle the result from a ValueTask in the same way the exception is handled

Sharing my code

```cs

static async void HandleSafeRun<TException, TResult>(ValueTask<TResult> valueTask, bool continueOnCapturedContext, 
        Action<TException> onException, 
        Action<TResult> onSuccess,
        bool rethrow = false) where TException : Exception
    {
        try
        {
            TResult result = await valueTask.ConfigureAwait(continueOnCapturedContext);
            HandleResult(result, onSuccess);
        }
        catch (TException ex) when (onException is not null)
        {
            HandleException(ex, onException);

            if (rethrow)
                throw;
        }
    }

    static void HandleResult<TResult>(in TResult result, in Action<TResult> onSuccess)
    {
        onSuccess?.Invoke(result);
    }

```

**Comment by TheCodeTraveler** [comment]
Sure! I think this new feature makes a lot of sense 💯

I'd probably add it to the `Task` APIs too, along with the `ValueTask` APIs. Just out of curiousity - is there any reason why you specifically called out `ValueTask` APIs? I.e. Is there an unforseen downside that I haven't yet thought about for adding this to the `Task` APIs too?

**Reply by stefanoIT** [reply]
Thanks for your reply, Brandon.
The only reason I implemented the method with ValueTask is that I'm using it, instead of Task<T> to reduce as much as I can the memory usage, using pooled objects results, in a performance-critical environment.
I don't see any downside about adding to Task methods as well, there are situations where it is definitely better suited! Even more, considering Net8+ `ConfigureAwaitOptions `

I should have been more accurate in my suggestion. I am glad you got the point!
